### PR TITLE
[#2] Add Markdown parsing for posts

### PR DESCRIPTION
**Pull request by nmertaugh**
See issue #2 for details

**Because:**

* We want to add Markdown support for authored posts

**This PR:**

* Adds markdown parsing of post bodies 
* Offers an ajax-based preview of markdown bodies during post authoring/editing

### CODE REVIEW

**Comment by nmertaugh** [comment]
@kirkland - lemme proof read this dude this afternoon before you hop on it; pounded through most of it stream-of-consciousness style and I want to give it a polishing.

> Then why did you submit the PR if it's not ready for review?

Because it makes me feel accomplished! Even though perhaps I'm not... :cry: But nearly!

**Comment by nmertaugh** [comment]
@kirkland - have at it buddy. We'll probably want to clean up the coffee script but I'll wait for you to weigh in.

Thanks buddy!

File: app/controllers/markdown_previews_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class MarkdownPreviewsController < ApplicationController
```
Comment (kirkland) [inline app/controllers/markdown_previews_controller.rb]: Should we require an admin login for this controller?

File: app/views/posts/show.haml [hunk 2]
```diff
@@ -1,8 +1,14 @@
-%h2= @post.title
+.container
+  %h2
+    = @post.title
+    %small.pull-right By #{@post.author.name}, #{l @post.created_at, format: :long}
 
-%span.pull-right By #{@post.author.name}, #{l @post.created_at, format: :long}
+  - if signed_in? && current_user.admin?
```
Comment (kirkland) [inline app/views/posts/show.haml]: Maybe add a helper for this combination? Something like `admin_user?`?

File: spec/helpers/application_helper_spec.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+describe ApplicationHelper, type: :controller do
+  include ApplicationHelper
+
+  describe '#markdown' do
+    it 'returns a markdown-parsed version of passed string' do
+      parser_stub = double(:parser, render: 'parsed')
+      allow(Redcarpet::Markdown).to receive(:new).and_return(parser_stub)
+
+      expect(markdown('unparsed')).to eq('parsed')
```
Comment (kirkland) [inline spec/helpers/application_helper_spec.rb]: I don't want to get hung up on inconsequential things, so feel free to disregard this. I think this test is too implementation-dependent (like, if we change our Markdown-parsing gem, we'll need to change this test). Perhaps it would be better to input some trivial text and test that its actual translation to HTML. But, then again, I haven’t tried that out, and that might just as implementation-dependent. So, whatever. Like I said, this is inconsequential.

File: spec/controllers/markdown_previews_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      it 'returns error message' do
+        post :create, invalid_params
+
+        expect(response.status).to eq(200)
+        expect(parsed_response[:error]).to eq(%(Unparsed_markdown can't be blank))
+      end
+    end
+  end
+
+  def invalid_params
```
Comment (kirkland) [inline spec/controllers/markdown_previews_controller_spec.rb]: Make these methods private?

File: app/assets/stylesheets/application.scss [hunk 5]
```diff
@@ -1,7 +1,7 @@
 /*
- *= require ./bootstrap_base
+ *= require bootstrap_base
  *= require font-awesome
  */
 
-@import "bourbon";
-@import 'common/index'
+@import 'bourbon';
```
Comment (kirkland) [inline app/assets/stylesheets/application.scss]: Are the semicolons necessary?

File: app/assets/javascripts/posts.coffee [hunk 6]
```diff
@@ -1,8 +1,49 @@
 ready = ->
+  getPreviewFor = (unparsedMarkdown) ->
+    $.ajax
+      data: { unparsed_markdown: unparsedMarkdown }
+      dataType: 'json'
+      type: 'POST'
+      url: '/markdown_previews'
+      success: (data) ->
+        swapPreLoadAlertFor data.parsed_markdown
+        return false
```
Comment (kirkland) [inline app/assets/javascripts/posts.coffee]: Use implicit return?

File: app/assets/javascripts/posts.coffee [hunk 7]
```diff
@@ -1,8 +1,49 @@
...
     $('form').find('input[type=text]:visible:first').focus()
 
+  initializeToggleableTabs = () ->
+    $('.nav-tabs').on 'click', 'a[data-toggle=tab]', (event) ->
+      tabTarget = $(event.target).attr('href')
+
+      if tabTarget == '#markdown-preview'
+        rawPostBody = $('#new-post-form form textarea:first').val()
+
+        if rawPostBody
```
Comment (kirkland) [inline app/assets/javascripts/posts.coffee]: So we check for blankness both here and the controller? Seems redundant to me… I think we can do without  the server-side check.

File: app/assets/javascripts/posts.coffee [hunk 8]
```diff
@@ -1,8 +1,49 @@
...
+      if tabTarget == '#new-post-form'
+        $('#new-post-form form textarea:first').focus()
+
+  showPreLoadAlert = () ->
+    $('#markdown-preview .post-load').hide()
+    $('#markdown-preview .pre-load').show()
+
+  swapPreLoadAlertFor = (body) ->
+    $('#markdown-preview .pre-load').hide()
+    $('#markdown-preview .post-load').html(body).removeClass('hide').show()
```
Comment (kirkland) [inline app/assets/javascripts/posts.coffee]: Is the `.show()` necessary here, as you’re already removing the `hide` class?

File: app/views/layouts/application.haml [hunk 9]
```diff
@@ -5,6 +5,7 @@
     %meta{ name: 'viewport', content: 'width=device-width, initial-scale=1' }
     = stylesheet_link_tag font_url
     = stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track' => true
+    = stylesheet_link_tag params['controller'], media: 'all', 'data-turbolinks-track' => true
```
Comment (kirkland) [inline app/views/layouts/application.haml]: When I request the home page, the page tries to include home.css, which doesn't exist.

**Comment by kirkland** [comment]
Looks great, and as a bonus, it even works! My comments are pretty minor.

**Comment by nmertaugh** [comment]
Thanks @kirkland! I'll address your comments.

File: app/controllers/markdown_previews_controller.rb (hunk 1 above)
Comment (nmertaugh) [inline app/controllers/markdown_previews_controller.rb]: Sure!

File: app/views/posts/show.haml (hunk 2 above)
Comment (nmertaugh) [inline app/views/posts/show.haml]: Yeah, we surely should.

File: spec/helpers/application_helper_spec.rb (hunk 3 above)
Comment (nmertaugh) [inline spec/helpers/application_helper_spec.rb]: Na, it's definitely testing implementation, but I tend to err on that side because we don't want to test the 'red_carpet` gem, ya know? So we use this gem, and we've outsourced all the parsing behavior to it, so we just want to test that we called it.

That's one argument. The other is yours, which tests some gem functionality but keeps the test delightfully ignorant of implementation. I'm going to side with you on this buddy, I just can't resist the opportunity to pull out implementation testing!

File: spec/controllers/markdown_previews_controller_spec.rb (hunk 4 above)
Comment (nmertaugh) [inline spec/controllers/markdown_previews_controller_spec.rb]: Oh, sure. I never do private methods in specs but I don't have a reason for that. I guess they _should_ be private.

File: app/assets/stylesheets/application.scss (hunk 5 above)
Comment (nmertaugh) [inline app/assets/stylesheets/application.scss]: No but I thought they were preferred in `scss`, no? I'm happy to ditch them, they're ugly.

Comment (nmertaugh) [inline app/assets/stylesheets/application.scss]: Actually, we use semi-colons on every `@import` line, so if we want to stop doing that it should be in a different commit and it should sweep the whole project.

File: app/assets/javascripts/posts.coffee (hunk 6 above)
Comment (nmertaugh) [inline app/assets/javascripts/posts.coffee]: But the implicit return isn't `false`. That said, I don't think we care about what we return here, there's no event propagation for the ajax call.

File: app/assets/javascripts/posts.coffee (hunk 7 above)
Comment (nmertaugh) [inline app/assets/javascripts/posts.coffee]: I disagree, but not emphatically. Here's why I disagree though. I think they serve different purposes. Client side, it serves to save an unnecessary ajax call when the body is blank. Server side it serves to protect against `POST`s made outside the web app. In this particular case it isn't critical, but in general, I think it's a little bit loose to have your controller rely on the client to do anything in particular.

File: app/assets/javascripts/posts.coffee (hunk 8 above)
Comment (nmertaugh) [inline app/assets/javascripts/posts.coffee]: Probably not :-)

File: app/views/layouts/application.haml (hunk 9 above)
Comment (nmertaugh) [inline app/views/layouts/application.haml]: True. I'll put a gate on it.

**Comment by nmertaugh** [comment]
Back to you @kirkland!

File: spec/helpers/application_helper_spec.rb (hunk 3 above)
Comment (kirkland) [inline spec/helpers/application_helper_spec.rb]: Good point on not testing the red_carpet gem.

File: spec/controllers/markdown_previews_controller_spec.rb (hunk 4 above)
Comment (kirkland) [inline spec/controllers/markdown_previews_controller_spec.rb]: Yeah, it does feel kind of pointless, since we're not reusing this code in other files.

File: app/assets/stylesheets/application.scss (hunk 5 above)
Comment (kirkland) [inline app/assets/stylesheets/application.scss]: :+1:

File: app/assets/javascripts/posts.coffee (hunk 6 above)
Comment (kirkland) [inline app/assets/javascripts/posts.coffee]: Oh, I mean change `return false` to `false`, if that's syntactically possible.

File: app/assets/javascripts/posts.coffee (hunk 7 above)
Comment (kirkland) [inline app/assets/javascripts/posts.coffee]: Sure.

File: app/views/posts/index.haml [hunk 10]
```diff
@@ -1,5 +1,9 @@
-- if @posts.any?
-  - @posts.each do |post|
-    %h4= link_to post.title, post_path(post)
-- else
-  %h2.text-center There are no posts written yet
+.container
+  - if signed_in? && current_user.admin?
```
Comment (kirkland) [inline app/views/posts/index.haml]: Use `admin_user?` here.

File: app/views/posts/show.haml (hunk 2 above)
Comment (kirkland) [inline app/views/posts/show.haml]: Also use `admin_user?` here.

File: spec/controllers/posts_controller_spec.rb [hunk 11]
```diff
@@ -1,4 +1,5 @@
 require 'spec_helper'
+include ControllerHelper
```
Comment (kirkland) [inline spec/controllers/posts_controller_spec.rb]: I think you'll want this `include` to be inside of the `describe` block, because out here, it's going to include the module in `Object`.

**Comment by kirkland** [comment]
A few small comments, but no further review required. Merge away!

File: app/views/posts/index.haml (hunk 10 above)
Comment (nmertaugh) [inline app/views/posts/index.haml]: Oh nice, I set up the helper and then didn't use it.... slick....

File: spec/controllers/posts_controller_spec.rb (hunk 11 above)
Comment (nmertaugh) [inline spec/controllers/posts_controller_spec.rb]: yup
